src/cpp/ext/gsm/metadata_exchange.cc

```diff
@@ -147,77 +157,81 @@ class LocalLabelsIterable : public LabelsIterable {
-    std::string decoded_metadata;
-    bool metadata_decoded = absl::Base64Unescape(
-        remote_metadata.as_string_view(), &decoded_metadata);
-    if (metadata_decoded) {
-      struct_pb_ = google_protobuf_Struct_parse(
-          decoded_metadata.c_str(), decoded_metadata.size(), arena_.ptr());
-      type_ = GetStringValueFromUpbStruct(struct_pb_, kMetadataExchangeTypeKey,
-                                          arena_.ptr());
-    }
-  }
+  explicit PeerLabelsIterable(grpc_core::Slice remote_metadata)
+      : metadata_(std::move(remote_metadata)) {}
 
   absl::optional<std::pair<absl::string_view, absl::string_view>> Next()
       override {
-    if (struct_pb_ == nullptr) {
+    auto& struct_pb = GetDecodedMetadata();
+    if (struct_pb.struct_pb == nullptr) {
       return absl::nullopt;
     }
-    if (pos_++ == 0) {
-      return std::make_pair(kPeerTypeAttribute, type_);
+    if (++pos_ == 1) {
+      return std::make_pair(kPeerTypeAttribute,
+                            GetStringValueFromUpbStruct(
+                                struct_pb.struct_pb, kMetadataExchangeTypeKey,
+                                struct_pb.arena.ptr()));
     }
     // Only handle GKE type for now.
-    if (type_ != kGkeType) {
-      return absl::nullopt;
-    }
-    switch (pos_ - 1) {
-      case 1:
-        return std::make_pair(
-            kPeerPodNameAttribute,
-            GetStringValueFromUpbStruct(struct_pb_, kMetadataExchangePodNameKey,
-                                        arena_.ptr()));
-      case 2:
-        return std::make_pair(
-            kPeerContainerNameAttribute,
-            GetStringValueFromUpbStruct(
-                struct_pb_, kMetadataExchangeContainerNameKey, arena_.ptr()));
-      case 3:
-        return std::make_pair(
-            kPeerNamespaceNameAttribute,
-            GetStringValueFromUpbStruct(
-                struct_pb_, kMetadataExchangeNamespaceNameKey, arena_.ptr()));
-      case 4:
-        return std::make_pair(
-            kPeerClusterNameAttribute,
-            GetStringValueFromUpbStruct(
-                struct_pb_, kMetadataExchangeClusterNameKey, arena_.ptr()));
-      case 5:
-        return std::make_pair(
-            kPeerLocationAttribute,
-            GetStringValueFromUpbStruct(
-                struct_pb_, kMetadataExchangeLocationKey, arena_.ptr()));
-      case 6:
-        return std::make_pair(
-            kPeerProjectIdAttribute,
-            GetStringValueFromUpbStruct(
-                struct_pb_, kMetadataExchangeProjectIdKey, arena_.ptr()));
-      case 7:
-        return std::make_pair(
-            kPeerCanonicalServiceAttribute,
-            GetStringValueFromUpbStruct(struct_pb_,
-                                        kMetadataExchangeCanonicalServiceKey,
-                                        arena_.ptr()));
-      default:
+    switch (type_) {
+      case GcpResourceType::kGke:
+        switch (pos_) {
```

Comment:
Consider using a static array here:

```
struct GkeAttribute {
  absl::string_view otel_attribute;
  absl::string_view metadata_attribute;
};
static const std::array<GkeAttribute, 7> kGkeAttributeList = {
  { kPeerPodNameAttribute, kMetadataExchangePodNameKey },
  { kPeerContainerNameAttribute, kMetadataExchangeContainerNameKey },
  // ...
};
```

Then the code here becomes:

```
if (pos_ - 1 >= kGkeAttributeList.size()) return absl::nullopt;
const GkeAttribute& attr = kGkeAttributeList[pos_ - 1];
return std::make_pair(
    attr.otel_attribute,
    GetStringValueFromUpbStruct(
        struct_pb.struct_pb,
        attr.metadata_attribute,
        struct_pb.arena.ptr()));
```

Also, on line 237, you can change the hard-coded `8` to `kGkeAttributeList.size() + 1`.

With this approach, if we ever add new labels here, there will be only one place to update the code.

Reply:
done